capsules/src/bmp280.rs

```diff
@@ -0,0 +1,378 @@
+    Error,
+}
+
+/// Complies with the reading and writing protocol used by the sensor.
+struct I2cWrapper<'a> {
+    i2c: &'a dyn i2c::I2CDevice,
+}
+
+impl<'a> I2cWrapper<'a> {
+    fn write<const COUNT: usize>(
+        &self,
+        buffer: &'static mut [u8],
+        addr: Register,
+        data: [u8; COUNT],
+    ) {
+        buffer[0] = addr as u8;
+        for (i, d) in data.iter().enumerate() {
+            buffer[i + 1] = *d;
+        }
+        // TODO: propagate errors
+        self.i2c.enable();
+        self.i2c.write(buffer, COUNT as u8 + 1).unwrap();
+    }
+
+    /// Requests a read into buffer.
+    /// Parse the result using `parse_read`.
+    fn read(&self, buffer: &'static mut [u8], addr: Register, count: u8) {
+        buffer[0] = addr as u8;
+        // TODO: propagate errors
+        self.i2c.enable();
+        self.i2c.write_read(buffer, 1, count).unwrap();
+    }
+
+    fn disable(&self) {
+        self.i2c.disable()
+    }
+
+    fn parse_read(buffer: &[u8], count: u8) -> &[u8] {
+        &buffer[..(count as usize)]
+    }
+}
+
+pub struct Bmp280<'a, A: Alarm<'a>> {
+    i2c: I2cWrapper<'a>,
+    temperature_client: OptionalCell<&'a dyn hil::sensors::TemperatureClient>,
+    // This might be better as a `RefCell`,
+    // because `State` is multiple bytes due to the `CalibrationData`.
+    // `Cell` requires Copy, which might get expensive, while `RefCell` doesn't.
+    // It's probably not a good idea to split `CalibrationData`
+    // into a separate place, because it will make state more duplicated.
+    state: Cell<State>,
+    /// Stores i2c commands
+    buffer: TakeCell<'static, [u8]>,
+    /// Needed to wait for readout completion, which can take milliseconds.
+    /// It's possible to implement this without an alarm with busy polling, but that's wasteful.
+    alarm: &'a A,
+}
+
+impl<'a, A: Alarm<'a>> Bmp280<'a, A> {
+    pub fn new(i2c: &'a dyn i2c::I2CDevice, buffer: &'static mut [u8], alarm: &'a A) -> Self {
+        Self {
+            i2c: I2cWrapper { i2c },
+            temperature_client: OptionalCell::empty(),
+            state: Cell::new(State::Uninitialized),
+            buffer: TakeCell::new(buffer),
+            alarm: alarm,
+        }
+    }
+
+    /// Resets the device and brings it into a known state.
+    pub fn begin_reset(&self) -> Result<(), ErrorCode> {
+        self.buffer.take().map_or_else(
+            || panic!("BMP280 No buffer available!"),
+            |buffer| match self.state.get() {
+                State::Uninitialized | State::Error => {
+                    self.i2c.read(buffer, Register::ID, 1);
+                    self.state.set(State::InitId);
+                    Ok(())
+                }
+                State::MismatchedDevice => Err(ErrorCode::NODEVICE),
+                _ => Err(ErrorCode::ALREADY),
+            },
+        )
+    }
+
+    pub fn read_temperature(&self) -> Result<(), ErrorCode> {
+        use State::*;
```

Comment:
We prohibit using `use...*` in Tock, please just use `State::`.

Reply:
Done.